### ISSUE TITLE
Issue with deepof.visuals.animate_skeleton when using custom labels

### ISSUE THREAD

**Issue by guodong0912** [mlfpm/deepof#52]
Dear DeepOF team,

First of all, thank you so much for developing such an amazing tool!

I am working on an experiment using custom labels as follows:

deepof_custom = { 
    "Nose": ["Head"],
    "Head": ["RightHindPaw", "LeftHindPaw"],
    "RightHindPaw": ["TTI"],
    "LeftHindPaw": ["TTI"],
}

My project setup is as follows:
base_path = "/mnt/c/Users/dell/Desktop/Refine/"
project_path = base_path
video_path = base_path + "Videos"
table_path = base_path + "Tables"

my_deepof_project_raw = deepof.data.Project(
    project_path=project_path,
    video_path=video_path,
    table_path=table_path,
    project_name="deepof_try1",
    arena='polygonal-manual',
    animal_ids=["Stress", "Agg"],
    table_format="csv",
    video_format=".avi",
    bodypart_graph=deepof_custom,
    video_scale=280,
    iterative_imputation="partial",  # "full",
    smooth_alpha=1,
    exp_conditions=None,
)
According to the custom label documentation, I was able to obtain the correct skeleton graph (as shown below). 
![image](https://github.com/user-attachments/assets/1ad93d41-45d1-4c18-8652-ccc1ec9437f2)

However, when I call deepof.visuals.animate_skeleton, I encounter an error. It seems that while I can pass animal_ids (such as Stress), the default keypoints definition is still being used, which results in a KeyError: 'Stress_Left_ear'.
**Case 1**
video = deepof.visuals.animate_skeleton(
    my_deepof_project,
    embedding=embeddings,
    cluster_assignments=soft_counts,
    experiment_id="A67-DAY1",
    frame_limit=150,
    selected_cluster=1,
    dpi=40,
    center="arena",
)

html = display.HTML(video)
display.display(html)
plt.close()
![image](https://github.com/user-attachments/assets/5c98b1bf-c22a-4f88-8852-56067e267d0b)

**Case 2**
from IPython import display

video = deepof.visuals.animate_skeleton(
    my_deepof_project,
    experiment_id="A67-DAY1",
    frame_limit=500,
    dpi=60,
)

html = display.HTML(video)
display.display(html)
plt.close()
![image](https://github.com/user-attachments/assets/7da7e4f0-0a85-4c3e-b371-50423c62f221)

Could you please guide me on how to handle this issue?

Thank you in advance!

Best regards,
guodong

**Comment by guodong0912** [comment]
Sorry for accidentally closing the issue😭

**Comment by NoCreativeIdeaForGoodUserName** [comment]
Hi guodong0912,

Unfortunately, I currently do not have access to my working PC and will only be able to look more into the square issue next week. But I already can give you some information:

1. The custom labelling option is not yet fully implemented in Deepof as some body-part names (e.g. "Nose", "Left_ear", "Right_ear" etc.) are hard coded in the in the legacy code. This means that using completely custom names will lead to you not being able to use some features of Deepof. You can circumvent this by changing the table headers accordingly to names deepof uses, see for instance the deepof-14 labelling schema:

```
                "deepof_14": {
                    "Nose": ["Left_ear", "Right_ear"],
                    "Spine_1": ["Center", "Left_ear", "Right_ear"],
                    "Center": ["Left_fhip", "Right_fhip", "Spine_2"],
                    "Spine_2": ["Left_bhip", "Right_bhip", "Tail_base"],
                    "Tail_base": ["Tail_1"],
                    "Tail_1": ["Tail_2"],
                    "Tail_2": ["Tail_tip"],
                },
```
 
2. Your custom labelling schema deviates significantly from what we usually use with deepof. If you like you can anonymously upload a test sample of your dataset (videos and matching tables) with the link below and I'll have a look at how to best handle your case next week. 

[Here](https://datashare.mpcdf.mpg.de/s/DEbgPN02vwQH1O7) you have an upload link you can use to anonymously provide a subset of your video and table files to deepof. This will allow me to check where exactly the error originates from (the files will only be visible to us (the deepof team) upon upload). Files will only be used to handle and resolve your problem except you explicitly allow us to use them for other purposes as well (e.g. to include them in our test code).

**Comment by guodong0912** [comment]
Hi NoCreativeIdeaForGoodUserName,

Thank you very much for your reply 😊. I have uploaded a test sample (including video and matching table). look forward to further communication next week! Wishing you a wonderful Christmas holiday 🎄🎉

**Comment by NoCreativeIdeaForGoodUserName** [comment]
Hi guodong0912,

Thank you, and I hope you also were able to enjoy the holidays.

I now had time to check out your data; thanks for uploading it. First: Where does the box come from? Since you have more than one mouse not only the distance / position information for the mice itself is relevant, but also the distance information between mice (i.e. how far are they apart). For this reason the mice are connected in the body part graph. Usually noses and tail_bases get connected between all mice. In your case however, while the Nose is named "Nose", there is no tail base. Hence the "tail bases" are defined as new nodes, resulting in that box.

![GraphWrong](https://github.com/user-attachments/assets/9b5bc00b-a462-4269-b7c8-1d38a7aeec37)

So ideally, it should look more like this in terms of connections.

![GraphRight](https://github.com/user-attachments/assets/7abcc847-1f7f-4157-9a8e-da26bfa479b6)

There is currently unfortunately not really a workaround for this, since, as mentioned, the custom labelling option is at the moment not yet fully implemented in Deepof. Another issue is that the "areas" for your mice are vastly different from anything Deepof usually works with and hence cannot be calculated (see the warning "It seems you're using a custom labelling scheme which is missing key body parts. You can proceed, but not all areas will be computed." during project creation).

In short, in the current version Deepof will probably not be able to adapt enough for your data. You may be able to make it work to some degree, but this would also carry the risk of inducing unexpected behaviors. We plan to rework the custom labelling section next, to allow users to define their body parts using a GUI, which should fix this problem. However, this is only planned after the next bigger release of 0.8, and it would also be part of another major update, so it may take a while.  

Best regards,
Patrick

**Comment by guodong0912** [comment]
Hi Patrick,

Thank you so much for identifying the issue! I’ve been puzzled by those automatically generated points for quite a while, and now I understand that they were created as key points for the "tail base." Actually, my data does contain "tail base," but I had labeled it as "TTI." Once I renamed it to "Tail_base," those confusing boxes disappeared, and the graph now looks much closer to the example you shared (it esembles an adorable crab 🦀).
![image](https://github.com/user-attachments/assets/bde20b5c-06ca-4110-bdd5-665c2803ea5e)

I also agree that continuing with this setup for analysis might not be the best choice. Therefore, I have retrained my DeepLabCut model following the deep_14 setup (minus the three tail-end points) to hopefully achieve more consistent and meaningful results.

Lastly, I want to say that DeepOF is an amazing tool, and I truly look forward to the upcoming GUI and enhanced custom labelling version. It will undoubtedly be a huge help for users like me who have limited programming experience and simpler needs.

Best wishes,
guodong

**Comment by NoCreativeIdeaForGoodUserName** [comment]
Great to hear that you like our software. 

The pre-made deepof_11 graph is essentially deepof_14 but without the tail body parts. If you name your bodyparts using the same names as in the deepof_11  schema below, you should be good. 

If there are any more issues you can let me now. I'll close this issue for now.

```
"deepof_11": {
    "Nose": ["Left_ear", "Right_ear"],
    "Spine_1": ["Center", "Left_ear", "Right_ear"],
    "Center": ["Left_fhip", "Right_fhip", "Spine_2"],
    "Spine_2": ["Left_bhip", "Right_bhip", "Tail_base"],
},
```

Best regards,
Patrick
